src/transformers/models/superglue/modeling_superglue.py

```diff
@@ -0,0 +1,845 @@
+
+        return output
+
+
+class SuperGlueAttentionalPropagation(nn.Module):
+    def __init__(
+        self,
+        config: SuperGlueConfig,
+    ):
+        super().__init__()
+        self.descriptor_dim = config.descriptor_dim
+        self.num_heads = config.num_heads
+        self.attention = SuperGlueMultiHeadAttention(config)
+        self.mlp = SuperGlueMultiLayerPerceptron(
+            config,
+            [self.descriptor_dim * 2, self.descriptor_dim * 2, self.descriptor_dim],
+        )
+        nn.init.constant_(self.mlp.layers[-1].bias, 0.0)
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        source: torch.Tensor,
+        output_attentions: bool = False,
+        output_hidden_states: bool = False,
+    ) -> tuple:
+        attention_outputs = self.attention(x, source, source, output_attentions=output_attentions)
+        output = attention_outputs[0]
+        attention = attention_outputs[1:]
+
+        output = torch.cat([x, output], dim=1)
+        layer_outputs = self.mlp(output, output_hidden_states=output_hidden_states)
+
+        last_hidden_state = layer_outputs[0]
+        hidden_states = layer_outputs[1]
+
+        return last_hidden_state, hidden_states, attention
+
+
+class SuperGlueAttentionalGNN(nn.Module):
+    def __init__(
+        self,
+        config: SuperGlueConfig,
+    ):
+        super().__init__()
+        self.descriptor_dim = config.descriptor_dim
+        self.num_heads = config.num_heads
+        self.layers_types = config.gnn_layers_types
+        self.num_layers = len(self.layers_types)
+        self.layers = nn.ModuleList(
+            [
+                SuperGlueAttentionalPropagation(
+                    config,
+                )
+                for _ in range(self.num_layers)
+            ]
+        )
+
+    def forward(
+        self,
+        descriptors_0: torch.Tensor,
+        descriptors_1: torch.Tensor,
+        output_attentions: bool = False,
+        output_hidden_states: Optional[bool] = False,
+    ) -> Tuple[torch.Tensor, torch.Tensor, Optional[tuple], Optional[tuple]]:
+        all_hidden_states = () if output_hidden_states else None
+        all_attentions = () if output_attentions else None
+
+        if output_hidden_states:
+            same_num_keypoints = descriptors_0.shape[-1] == descriptors_1.shape[-1]
+            if not same_num_keypoints:
+                max_num_keypoints = max(descriptors_0.shape[-1], descriptors_1.shape[-1])
+                new_hidden_state = torch.zeros((2, self.descriptor_dim, max_num_keypoints)).to(descriptors_0.device)
+                new_hidden_state[:, :, : descriptors_0.shape[-1]] = descriptors_0
+                new_hidden_state[:, :, : descriptors_1.shape[-1]] = descriptors_1
+            else:
+                new_hidden_state = torch.cat([descriptors_0, descriptors_1])
+            all_hidden_states = all_hidden_states + (new_hidden_state,)
+
+        for gnn_layer, type in zip(self.layers, self.layers_types):
```

Comment:
We can't use type here - it overloads the in-built `type` function

Reply:
Fixed in [this commit](https://github.com/huggingface/transformers/pull/29886/commits/601da1cf0756cef7339d4dd33faeefbfcd073ac9)